sklearn/base.py

```diff
@@ -342,6 +344,71 @@ def _get_tags(self):
                 collected_tags.update(more_tags)
         return collected_tags
 
+    def _check_n_features(self, X, reset):
+        """Set the `n_features_in_` attribute, or check against it.
+
+        Parameters
+        ----------
+        X : {ndarray, sparse matrix} of shape (n_samples, n_features)
+            The input samples.
+        reset : bool
+            If True, the `n_features_in_` attribute is set to `X.shape[1]`.
+            Else, the attribute must already exist and the function checks
+            that it is equal to `X.shape[1]`.
+        """
+        n_features = X.shape[1]
+
+        if reset:
+            self.n_features_in_ = n_features
+        else:
+            if not hasattr(self, 'n_features_in_'):
+                raise RuntimeError(
+                    "The reset parameter is False but there is no "
+                    "n_features_in_ attribute. Is this estimator fitted?"
+                )
+            if n_features != self.n_features_in_:
+                raise ValueError(
+                    'X has {} features, but this {} is expecting {} features '
+                    'as input.'.format(n_features, self.__class__.__name__,
+                                       self.n_features_in_)
+                )
+
+    def _validate_data(self, X, y=None, reset=True, **check_params):
+        """Validate input data and set or check the `n_features_in_` attribute.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix, dataframe} of shape \
+                (n_samples, n_features)
+            The input samples.
+        y : array-like of shape (n_samples,), default=None
+            The targets. If None, `check_array` is called on `X` and
+            `check_X_y` is called otherwise.
```

Comment:
Previously, check_X_y would enforce that y was not None in many estimators. What happens if the user passes y=None now (e.g. through a Pipeline where y is not a required parameter)?

Reply:
I think the implication is that now this will break in tuple unpacking.

Reply:
I think you're correct Joel. 

Calls to `X, y = check_X_y(X, y)` have been replaced by `X, y = _validate_data(X, y)`

Calling `fit(X, y=None)` will now go through `check_array` instead of `check_X_y` because `y` is None (see code below), and only `X` will be returned instead of `X, y`.

So instead of the nicer error in `check_X_y` ("y cannot be None"), we now have a tuple unpacking error.

I'm not sure what to do here, apart e.g. replacing `check_X_y(X, y)` to `X, y = _validate_data(X, y, force_y_not_none=True)`

Reply:
+1 overall but rather than `force_y_not_none`, may I suggest I would `y_ignored=True/False` or  `y_required=True/False`?

The "force" terminology makes me think that the validation method can automatically convert `y` into something that is not `None` while here it will raise a `ValueError` if `y` is `None`

Reply:
Either use estimator tag or overriding of this method to specify that y is required on fit?

Reply:
It could be weird to start using estimator tags for this use-case. Up to now, we used estimator tags to run some specific common tests. Here, I have the impression that we want to use the tag such the estimator will change is internal validation, isn't it?

Reply:
IMO, passing a parameter will be more explicit there.

Reply:
Another solution is to force `y` to be passed, i.e. change the signature to 

`def _validate_data(self, X, y, reset=True, **check_params)`

and force every caller (even unsupervised estimators) to explicitly pass `y`. I think this is the best option since even unsupervised estimators are required by convention to define `def fit(self, X, y=None)`.

WDYT?

Reply:
> It could be weird to start using estimator tags for this use-case. Up to now, we used estimator tags to run some specific common tests.

Inheritance then?

How does forcing the caller to pass y help us check when the user has indirectly provided null y? Do you rather mean that _validate_data should always return an X, y pair? That would be good in some ways. It would produce a better error message than tuple unpacking

Reply:
Indeed my solution is useless on its own, we would still need the `y_required` parameter.

I guess inheritance could almost work, i.e. using our current Mixins, but how do we deal with meta-estimators?

-----

I feel like a tag could work here. We can define a `requires_y` tag which defaults to True and which is overriden to False in OutlierMixin, [Bi]ClusterMixin, and TransformerMixin.

For the meta-estimators, we can have adhoc handling, e.g. for pipelines that would be something like `tags['requires_y'] = any(step.get_tags['requires_y'] for step in self.steps)`

That's not a trivial change though and it should be done in another PR which should be merged first.

Reply:
> It could be weird to start using estimator tags for this use-case. Up to now, we used estimator tags to run some specific common tests. Here, I have the impression that we want to use the tag such the estimator will change is internal validation, isn't it?

I don't think the proposal above would change the internal validation. With my proposal above, we can add a common check that makes sure a proper error is raised when requires_y is True but None is passed.

My main concerns for using estimator tags outside of estimator checks was that it was impossible to override them in child classes more than once (fixed now), and that it required deprecation cycle / warnings to give time to third party developer to adapt (https://github.com/scikit-learn/scikit-learn/pull/13565#issuecomment-480294329). But that's a non-issue there: we're already introducing a common check with warnings for 2 cycles, so we might as well have 2 of them

Reply:
> It could be weird to start using estimator tags for this use-case. Up to now, we used estimator tags to run some specific common tests.

> Inheritance then?

I think it would be good if there was a programmatic way to figure out if an estimator requires y (cc @thomasjpfan). I think it's a valid question whether we want to use estimator tags for that. While it would be the first time we use estimator_tags in an estimator, I think it's a good use. Right now, we actually use ``_estimator_type`` when we call ``check_cv``. While ``_estimator_type`` is not an estimator tag, I think it actually should be.

Whether we want to make this decision a prereq for this PR is another question.
So maybe just always returning ``X, y`` is the simpler solution? I would be ok with either.

I opened #16468 for tracking adding a tag on requiring y for fitting.

Reply:
The only concern I have with returning both `X, y` is that this is a design we would not have come up with if we didn't encounter this issue. So it's more of a workaround than a deliberate design choice. Not strongly against it though